### DISCUSSION TITLE
Pass MQTT connection username and client_id to AMQP 0-9-1 message headers

### DISCUSSION THREAD

**Discussion by getlarge** [rabbitmq/rabbitmq-server#11632] [Other] [answered]
## Context

A system containing a mix of MQTT and AMQP consumers and producers.
We are using RabbitMQ v3.12.13.

## Problem

No information about the MQTT client is being passed to the AMQP message properties and fields.

It would be very helpful to access the client ID and username when our AMQP consumers process messages from MQTT producers to identify the producer.
I imagine the information is available since the RabbitMQ authorization methods (check resource and check topic) receive the username.
I suppose there is a reason why the information is not transferred in the AMQP payload during [`deliver_to_queues`](https://github.com/rabbitmq/rabbitmq-server/blob/v3.12.13/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1068)?

### Describe the solution you'd like

In the latest [publish_to_queues](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1534), the state contains the username, could we imagine setting an AMQP header containing the `username` and the `client_id`?

Do I suppose correctly that #10050 and #10051 could help achieve this?

## Potential solutions

- Use MQTT v5.0 and include the user ID in the message properties
- Use a custom MQTT plugin that includes the user ID in the message properties
- Pass the MQTT client ID and user ID in the payload and/or topic

All these solutions are quite complex in our system since we don't have control over the MQTT clients' code.

**Comment by michaelklishin** [comment]
3.12.x is out of [community support](https://www.rabbitmq.com/release-information) and the MQTT plugin in 3.13.x was [substantially reworked](https://www.rabbitmq.com/blog/2023/03/21/native-mqtt) to support MQTT 5.0.

https://github.com/rabbitmq/rabbitmq-server/issues/10050 is not MQTT-specific.

**Reply by getlarge** [reply]
The problem remains with RabbitMQ 3.13.7, except that the solution could be implemented in the [mc:init](https://github.com/rabbitmq/rabbitmq-server/blob/v3.13.7/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1579-L1580), right?
Could we consider passing `state's cfg` and `auth_state` in the `mc:init` arguments?

**Reply by michaelklishin** [reply]
I'd start with a more fundamental question: do most users need MQTT publisher username to be passed to non-MQTT consumers? Not as far as I can tell.

Client ID as passed to authN/authZ backends (where I can see how it can be universally useful), as of https://github.com/rabbitmq/rabbitmq-server/issues/9323 it can be compared to a x.509 certificate-provided identity.

If consumers need some information from publishers, publishers should put it into the message (or message metadata, if the protocol allows for that).

RabbitMQ message translation should be tweaked only when there's potential to simplify the lives of a significant majority of users, based on real world evidence. I am not convinced that this is the case here.

**Reply by robertsLando** [reply]
Hi @michaelklishin , thanks for your answer. 

What I don't understand is if making this change has some drawbacks on RabbitMQ side or not, I'm open to help contributing to source code to integrate such feature if it's not harmful for RabbitMQ as it's something blocking for us on our company and I think it could be for others using RabbitMQ as MQTT broker, unfortunately adding client username to payloads isn't a valid option as it would increese each payload size in vain.

> The problem remains with RabbitMQ 3.13.7, except that the solution could be implemented in the [mc:init](https://github.com/rabbitmq/rabbitmq-server/blob/v3.13.7/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1579-L1580), right?
Could we consider passing state's cfg and auth_state in the mc:init arguments?

Is @getlarge statement correct?

**Reply by michaelklishin** [reply]
@robertsLando contributing the code is good and all but are you willing to maintain this feature down the road? Most contributors are not, so the RabbitMQ core team has to consider every behavior we accept as something we will have to maintain for years.

I hope I was very clear that this is not a commonly requested feature, neither it is an obvious way of integrating MQTT publishers and AMQP 0-9-1 consumers. So no, I'm afraid I do not see any evidence that other MQTT users do things the way you intend to.

We have a long history of accepting such "user-specific" features that start collecting dust once their original contributor decides to do things differently. Curiously they almost always have to do with MQTT, likely because it is not the most extensible or flexible protocol out there.

**Reply by michaelklishin** [reply]
@robertsLando 

> adding client username to payloads isn't a valid option as it would increese each payload size in vain

You do consider the payload increase for your apps but completely ignore the fact that this extra translation will waste CPU cycles and bandwidth (on the consumer side) for every RabbitMQ MQTT user, in every cluster with MQTT publishers, today and in the future, whether they need this information to be added to the message or not.

I will repeat myself but I do not see other RabbitMQ users who use MQTT raise the need for these fields — not present in the original message — to be added. Moving connection data to messages is not something that RabbitMQ does anywhere else, for any combination of protocols.

**Reply by robertsLando** [reply]
@michaelklishin I completely understand what you mean as I also maintain many open source libraries with large user base so I also have to deal with such requests. This is why I asked you if it is harmful for RabbitMQ or not, also about maintain such feature I'm completely open for that as I'm very active in OSS world but if it's something that could slow down other users I understand this is not good

**Reply by michaelklishin** [reply]
> Is @getlarge statement correct?

`mc:init/4` is not generally aware of connection details, so any connection-level metadata will have to be passed in as keys in a "message container environments", which is a protocol-specific map returned by `mc_env/0` on the linked line. Then those values can be used in `mc:init/4` to populate message metadata.

The license allows you to fork RabbitMQ, make the changes and run such a modified version. The core team or Broadcom will not support such a fork, of course.

**Reply by michaelklishin** [reply]
It is not harmful per se. It will carry a small but non-zero resource cost for every message published via MQTT, and a certain maintenance cost.

I have shared this discussion with the rest of our team to see if I'm "reading the room" correctly w.r.t. how many times something like that was requested, and what other engineers think.

**Reply by robertsLando** [reply]
@michaelklishin thanks for taking the time to discuss this with your team, let me know!

**Reply by michaelklishin** [reply]
Some team members would like to work on message interceptor support for 4.1.0 (which will ship around Q2 2025. Message interceptors is an abstraction that originates from RabbitMQ's original protocol implementation, AMQP 0-9-1, that allows plugins to hook into the handlers of specific protocol frames, and modify them as necessary.

Here is a popular and small example of such as plugin: https://github.com/rabbitmq/rabbitmq-message-timestamp

So far no one has expressed explicit support for a core change suggested above.

**Reply by michaelklishin** [reply]
In fact, since you @robertsLando have AMQP 0-9-1 consumers, perhaps a "consumer side" channel interceptor plugin would be an option today (with 3.13.x and the upcoming 4.0.x).

**Reply by getlarge** [reply]
Thanks for the suggestion @michaelklishin. Is there documentation or some more examples to help create such a plugin?

**Reply by michaelklishin** [reply]
The plugin above is less than 100 lines of code. There is similar plugin, `rabbitmq_node_stamp`, which is even smaller.

There's also an example plugin, `rabbitmq_metronome`.

**Reply by robertsLando** [reply]
I think we will go for that way, thanks again @michaelklishin 🙏🏼

**Reply by getlarge** [reply]
I just tried to create a `channel_interceptor` based on RabbitMQ 3.13.7, and unfortunately, it only intercepts messages published by AMQP producers.
The behaviour was excellent since I could use the `init` method to create an interceptor state based on the channel state (and extract usernames and tags).

I noticed the `message_interceptor` used in `rabbitmq_mqtt/src/rabbit_mqtt_processor.erl`, but we cannot extract `auth_state,` right?
Also, when trying to create a message interceptor and running the build, I received the error `behaviour rabbit_message_interceptor undefined`. 
Is this expected?
Is it possible that: `class_module(message_interceptor)       -> rabbit_message_interceptor;` is missing from `rabbit_common/src/rabbit_registry.erl`?

**Reply by robertsLando** [reply]
Just for reference:

https://github.com/rabbitmq/rabbitmq-server/issues/10051
https://groups.google.com/g/rabbitmq-users/c/gPJxxG2PGXM/m/EF2BDwpIAQAJ?pli=1

**Comment by ansd** [comment, marked as answer]
Passing the MQTT client ID of the publishing client to an AMQP 1.0 or AMQP 0.9.1 message of receiving clients will be possible in RabbitMQ 4.2 as explained in https://www.rabbitmq.com/docs/next/message-interceptors#mqtt-client-id and https://github.com/rabbitmq/rabbitmq-server/pull/13760
